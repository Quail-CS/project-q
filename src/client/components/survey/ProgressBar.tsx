/**
 * @module ProgressBar.tsx
 * @description Issue Selection and Survey Progress Bar.
 * Returns a bar with progress level determined by # of issues selected & answered
 */

import * as React from 'react';

interface ProgressBarProps {
  surveyPage: number,
  issuesCount: number,
}

interface progressBarFillProps {
  percentComplete: number,
}

const ProgressBarFill = (props: progressBarFillProps) => {
  const { percentComplete } = props;
  return (
    <div
      className="progress-bar-fill"
      style={{ width: `${percentComplete}%`}}
    >
      {percentComplete}%
    </div>
  )
}

// Accept Page Number and Issues Count
const ProgressBar = (props: ProgressBarProps) => {
  const { surveyPage, issuesCount } = props;

  // Helper function that returns a formatted whole percentage
  const percentComplete = Math.floor(((surveyPage + 1) / (issuesCount + 1)) * 100);
  

  return (
    <div className="progress-bar-container">
      <div className="progress-bar">
        <ProgressBarFill percentComplete={percentComplete}/>
      </div>
    </div>
  )
}

export default ProgressBar;

export { ProgressBarFill };
