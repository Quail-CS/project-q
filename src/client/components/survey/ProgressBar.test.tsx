import * as React from 'react';
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as Renderer from 'react-test-renderer';

import ProgressBar, { ProgressBarFill } from './ProgressBar';

Enzyme.configure({ adapter: new Adapter() });

// Declare constants for testing
const stubSurveyPage = 1;
const stubIssueCount = 4;
const stubPercentComplete = 40;

describe('Progress Bar Component: ', () => {
  it('matches the snapshot', () => {
    const tree: Renderer.ReactTestRendererJSON = Renderer.create(<ProgressBar surveyPage={stubSurveyPage} issuesCount={stubIssueCount}/>).toJSON() as Renderer.ReactTestRendererJSON;
    expect(tree).toMatchSnapshot();
  })
  it('renders a <div> with class "progress-bar-container"', () => {
    const progressBarWrapper: Enzyme.ShallowWrapper = Enzyme.shallow(<ProgressBar surveyPage={stubSurveyPage} issuesCount={stubIssueCount}/>);
    expect(progressBarWrapper.type()).toEqual('div');
    expect(progressBarWrapper.hasClass('progress-bar-container')).toEqual(true);
  })
  it('contains a <div> with class "progress-bar" which contains a ProgressBarFill with props', () => {
    const progressBarWrapper: Enzyme.ShallowWrapper = Enzyme.shallow(<ProgressBar surveyPage={stubSurveyPage} issuesCount={stubIssueCount}/>);
    expect(progressBarWrapper.find('.progress-bar')).toHaveLength(1);
    expect(progressBarWrapper.find('.progress-bar').type()).toEqual('div');
    expect(progressBarWrapper.find(ProgressBarFill)).toHaveLength(1);
    expect(progressBarWrapper.find(ProgressBarFill).props().percentComplete).toEqual(40);
  })
  it('AND ProgressBarFill has props equal to 100 * (surveyPage + 1 / issueCount + 1)', () => {
    const expectedPercentComplete = ((stubSurveyPage + 1)/(stubIssueCount + 1)) * 100;
    const progressBarWrapper: Enzyme.ShallowWrapper = Enzyme.shallow(<ProgressBar surveyPage={stubSurveyPage} issuesCount={stubIssueCount}/>);
    expect(progressBarWrapper.find(ProgressBarFill)).toHaveLength(1);
    expect(progressBarWrapper.find(ProgressBarFill).props().percentComplete).toEqual(expectedPercentComplete);
  })
});

describe('Progress Bar Fill Component: ', () => {
  it('matches the snapshot', () => {
    const tree: Renderer.ReactTestRendererJSON = Renderer.create(<ProgressBarFill percentComplete={stubPercentComplete} />).toJSON() as Renderer.ReactTestRendererJSON;
    expect(tree).toMatchSnapshot();
  })
  it('renders a <div> with class "progress-bar-fill"', () => {
    const progressBarWrapper: Enzyme.ShallowWrapper = Enzyme.shallow(<ProgressBarFill percentComplete={stubPercentComplete} />);
    expect(progressBarWrapper.type()).toEqual('div');
    expect(progressBarWrapper.hasClass('progress-bar-fill')).toEqual(true);
  })
  it('has width and text equal to percentComplete', () => {
    const progressBarWrapper: Enzyme.ShallowWrapper = Enzyme.shallow(<ProgressBarFill percentComplete={stubPercentComplete} />);
    expect(progressBarWrapper.text()).toEqual(`${stubPercentComplete}%`)
    expect(progressBarWrapper.prop('style')).toEqual({"width": `${stubPercentComplete}%`});
  })
});