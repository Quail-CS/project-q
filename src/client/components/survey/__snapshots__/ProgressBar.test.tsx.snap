// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress Bar Component:  matches the snapshot 1`] = `
<div
  className="progress-bar-container"
>
  <div
    className="progress-bar"
  >
    <div
      className="progress-bar-fill"
      style={
        Object {
          "width": "40%",
        }
      }
    >
      40
      %
    </div>
  </div>
</div>
`;

exports[`Progress Bar Fill Component:  matches the snapshot 1`] = `
<div
  className="progress-bar-fill"
  style={
    Object {
      "width": "40%",
    }
  }
>
  40
  %
</div>
`;
